kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/storage/PartitionLog.java

```diff
@@ -0,0 +1,257 @@
+    private String namespacePrefix;
+    private String fullPartitionName;
+    private EntryFormatter entryFormatter;
+    private Optional<TransactionCoordinator> transactionCoordinator;
+
+    public CompletableFuture<Long> appendRecords(final MemoryRecords records,
+                                                 final short version,
+                                                 final AppendRecordsContext appendRecordsContext) {
+        return append(records, version, appendRecordsContext);
+    }
+
+    /**
+     * Append this message to pulsar.
+     *
+     * @param records The log records to append
+     * @param version Inter-broker message protocol version
+     * @param appendRecordsContext See {@link AppendRecordsContext}
+     */
+    private CompletableFuture<Long> append(final MemoryRecords records,
+                                           final short version,
+                                           final AppendRecordsContext appendRecordsContext) {
+        CompletableFuture<Long> appendFuture = new CompletableFuture<>();
+        RequestStats requestStats = appendRecordsContext.getRequestStats();
+        KafkaTopicManager topicManager = appendRecordsContext.getTopicManager();
+        final long beforeRecordsProcess = time.nanoseconds();
+        try {
+            MemoryRecords validRecords = validateRecords(version, fullPartitionName, records);
+            validRecords.batches().forEach(batch->{
+                if (batch.sizeInBytes() > kafkaConfig.getMaxMessageSize()) {
+                    throw new RecordTooLargeException(String.format("Message batch size is %s "
+                                    + "in append to partition %s which exceeds the maximum configured size of %s .",
+                            batch.sizeInBytes(), topicPartition, kafkaConfig.getMaxMessageSize()));
+                }
+            });
+            // Append Message into pulsar
+            final CompletableFuture<Optional<PersistentTopic>> topicFuture =
+                    topicManager.getTopic(fullPartitionName);
+            if (topicFuture.isCompletedExceptionally()) {
+                topicFuture.exceptionally(e -> {
+                    appendFuture.completeExceptionally(e);
+                    return Optional.empty();
+                });
+                return appendFuture;
+            }
+            if (topicFuture.isDone() && !topicFuture.getNow(Optional.empty()).isPresent()) {
+                appendFuture.completeExceptionally(Errors.NOT_LEADER_FOR_PARTITION.exception());
+                return appendFuture;
+            }
+            final Consumer<Optional<PersistentTopic>> persistentTopicConsumer = persistentTopicOpt -> {
+                if (!persistentTopicOpt.isPresent()) {
+                    appendFuture.completeExceptionally(Errors.NOT_LEADER_FOR_PARTITION.exception());
+                    return;
+                }
+                // TODO: validateMessagesAndAssignOffsets here.
```

Comment:
`validateMessagesAndAssignOffsets` is called in `MixedKafkaEntryFormatter#encode`.

Reply:
In idempotent producer, we need to assign offset here first and we need to support all formatter.